packages/reward-root-submitter/reward_root_submitter/utils.py

```diff
@@ -0,0 +1,143 @@
+import json
+import re
+from typing import Any, Callable, Dict, TypedDict
+
+import pandas as pd
+import requests
+from cloudpathlib import AnyPath
+from web3 import Web3
+
+from .config import Config
+
+DEFAULT_SUBGRAPH_PAGINATE_SIZE = 100
+
+
+def safe_regex_group_search(regex, string, group):
+    """
+    Returns None in the case of a missing group
+    """
+    match = re.search(regex, string)
+    if match:
+        return match.group(group)
+    else:
+        return None
+
+
+class SubgraphQuery(TypedDict):
+    name: str
+    query: str
+
+
+def merkle_root_submissions_query(
+    skip: int, blockNumber_gt=0, **filters
+) -> SubgraphQuery:
+    query = """{{
+        merkleRootSubmissions(
+            skip: {skip},
+            orderBy: blockNumber,
+            where: {{ blockNumber_gt: {blockNumber_gt} }}
+        )
+        {{
+            rewardProgram {{
+                id
+            }}
+            paymentCycle
+        }}
+    }}""".format(
+        **{"skip": skip, "blockNumber_gt": blockNumber_gt, **filters}
+    )
+    return {"name": "merkleRootSubmissions", "query": query}
+
+
+def query_subgraph(
+    config: Config,
+    skip: int,
+    blockNumber_gt: int,
+    query: Callable[[int, int, Dict[str, Any]], SubgraphQuery],
+):
+    print(f"Skipping {skip}")
+    subgraph_url = config.subgraph_url
+    subgraph_query = query(skip, blockNumber_gt)
+    r = requests.post(
+        subgraph_url,
+        json={"query": subgraph_query["query"]},
+    )
+    if r.ok:
+        json_data = json.loads(r.text)
+        data = json_data["data"]
+        return data[subgraph_query["name"]]
+
+
+def get_roots_subgraph(config: Config, min_scan_block=0):
+    paginate_size = 100
+    res = []
+
+    i = 0
+    while c := query_subgraph(
```

Comment:
Looks good - I wondered about trying to avoid explicit pagination and update the block number we're filtering on but the end condition gets really awkward, I think this pagination is cleaner.